spec/lib/configuration.spec.js

```diff
@@ -0,0 +1,295 @@
+'use strict';
+
+const Configuration = require('../../lib/configuration').Configuration;
+const CLIOptionsMock = require('../mocks/cli-options');
+
+describe('the Configuration module', () => {
+  let cliOptionsMock;
+
+  beforeEach(() => {
+    cliOptionsMock = new CLIOptionsMock();
+    cliOptionsMock.attach();
+  });
+
+  afterEach(() => {
+    cliOptionsMock.detach();
+  });
+
+  it('overrides default options with customized options', () => {
+    let config = new Configuration({ fromString: true }, { fromString: false });
+    expect(config.getAllOptions().fromString).toBe(true);
+  });
+
+  describe('the getAllOptions() function', () => {
+    it('returns the entire options object', () => {
+      let options = {
+        rev: 'dev & prod',
+        minify: true,
+        inject: { dev: { value: true } }
+      };
+      let config = new Configuration({}, options);
+      expect(config.getAllOptions()).toEqual(options);
+    });
+  });
+
+  describe('the getValue() function', () => {
+    it('supports multi level options', () => {
+      let options = new Configuration({}, {
+        foo: {
+          bar: {
+            dev: {
+              value: 'someValue'
+            },
+            staging: {
+              value: 'someOtherValue'
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value).toBe('someValue');
+
+      options = new Configuration({}, {
+        foo: {
+          bar: {
+            'dev & prod': {
+              value: 'someValue'
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value).toBe('someValue');
+
+      options = new Configuration({}, {
+        foo: {
+          bar: {
+            dev: {
+              value: {
+                options: true
+              }
+            }
+          }
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar').value.options).toBe(true);
+
+      options = new Configuration({}, {
+        foo: {
+          bar: 'abcd'
+        }
+      }, 'dev');
+      expect(options.getValue('foo.bar')).toBe('abcd');
+    });
+
+    it('supports one level options', () => {
+      let options = new Configuration({}, { foo: 'someValue' }, 'dev');
+      expect(options.getValue('foo')).toBe('someValue');
```

Comment:
Yep, that's a bit tricky. I think it's OK to check whether or not the value is an object or a string, because only the code that uses the Configuration class knows what it could/should be